woocommerce/woocommerce-gutenberg-products-block
# Handle translations in lazy loaded files

Adds support for loading translations for lazy-loaded components.

`RequireChunkCallbackPlugin` is borrowed from Calypso.

On the blocks side, we use `addRequireChunkTranslationsHandler` to provide a list of Chunk IDs containing translations. `getTranslationChunkFileUrl` is a callback that maps the Chunk ID to a JSON filename in the WP_LANG dir. This consists of the textdomain, locale, and a hash for the script file itself.

When a file is imported by React, `addRequireChunkTranslationsHandler` will run and fetch the JSON file. If successful, we simply pass the body to `setLocaleData` so the translation is then performed by the `@wordpress/i18n` package.

Fixes #4019
Closes #4392

**Testing**

**Testing translations:**

1. Add the all products block to a page and save.
2. Ensure a product is on sale that's visible in this block.
3. Switch to a non US language with translations e.g. Brazillian Portuguese. 
4. View the page on the frontend with the all products block. Ensure "sale" is translated.

**Testing builds:**

Also, test builds (dev and production) because this required an update to Webpack 5x.

**Core plugin**

TBC - We need a way to test under core context because the domain and paths are different. I've added handling for this but it's not tested.

**Changelog**

> Fixed string translations within the All Products Block.
